### PR TITLE
Fix milestone deadline and date related problems

### LINKED ISSUE #32291: JavaScript Error: Cannot Parse Date Format on Version 1.21.11 During Page Change in Milestones
### Description:
During the page change in the Milestones section, a JavaScript error is encountered that states:

```js
JavaScript error: Cannot parse: 10000-01-0
```

This error is thrown from the file webcomponents.js at the following URL:
https://10.11.20.95:3000/assets/js/webcomponents.js?v=1.21.11 at line 1,55713.

### Steps to Reproduce:
1. Navigate to the Milestones section.
2. Attempt to change the page.
3. Observe the error in the browser console.

### Expected Behavior:
The page should change without any JavaScript errors.

### Actual Behavior:
A parsing error occurs, but the milestone list is loaded correctly.

### Recommendation:
Check the date formatting in the data being passed to the JavaScript function. Ensure it conforms to a valid date format that can be parsed successfully.

### Note:
Open the browser console for more details on the error message and stack trace.

### Gitea Version

1.21.11

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

image

### Git Version

_No response_

### Operating System

Ubuntu

### How are you running Gitea?

The Gitea is running on Ubuntu 20.04.3 as a systemd service.

### Database

MySQL/MariaDB

### PR DESCRIPTION
**Pull request by lunny**
Use zero instead of 9999-12-31 for deadline
Fix #32291

### CODE REVIEW

File: models/issues/milestone.go [hunk 1]
```diff
@@ -82,9 +84,18 @@ func (m *Milestone) BeforeUpdate() {
...
+	if m.DeadlineUnix == 0 {
+		return
+	}
+	// for legacy reasons, all years after 9000 are considered as no deadline
+	if m.DeadlineUnix.Year() > 9000 {
+		m.DeadlineUnix = 0
+		m.IsOverdue = false
+		if _, err := session.ID(m.ID).NoAutoTime().Cols("deadline_unix", "is_overdue").Update(m); err != nil {
+			log.Error("AfterLoad update legacy deadline failed: %v", err)
+		}
```
Comment (wolfogre) [inline models/issues/milestone.go]: Why not add a migration to convert all legacy deadlines to the new value in the DB instead of keeping this code?

Comment (lunny) [reply models/issues/milestone.go]: To make the backport easier.

File: models/issues/milestone.go [hunk 2]
```diff
@@ -82,9 +84,18 @@ func (m *Milestone) BeforeUpdate() {
...
 	m.NumOpenIssues = m.NumIssues - m.NumClosedIssues
-	if m.DeadlineUnix.Year() == 9999 {
+	if m.DeadlineUnix == 0 {
+		return
+	}
+	// for legacy reasons, all years after 9000 are considered as no deadline
+	if m.DeadlineUnix.Year() > 9000 {
+		m.DeadlineUnix = 0
+		m.IsOverdue = false
+		if _, err := session.ID(m.ID).NoAutoTime().Cols("deadline_unix", "is_overdue").Update(m); err != nil {
```
Comment (wxiaoguang) [inline models/issues/milestone.go]: Is it right to do so? Abusing the sessions would cause bugs in XORM, like this one  Fix db engine #32351

Comment (lunny) [reply models/issues/milestone.go]: It's right. There are tests in XORM side. I will take a look at another problem you mentioned above.

**Comment by wxiaoguang** [comment]
The fix doesn't look good to me.

IMO it should be done separately.

* For 1.23: use a migration to update database and remove all "9999-12-31" from code (or only leave a few places to handle API input).
* For 1.22: only fix the value used by frontend but no need to touch other code. Actually there could be a one-line frontend patch to make it work without any backend change.

**Comment by wxiaoguang** [comment]
Frontend fix:  Fix absolute-date #32375

File: models/issues/milestone.go (hunk 1 above)
Comment (lunny) [reply models/issues/milestone.go]: 6d0f84029ca6af991b8a161c4834688d07596df9

**Review by wolfogre — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
And:  Quick fix milestone deadline 9999 for 1.22 #32423

File: routers/web/repo/issue.go [hunk 3]
```diff
@@ -2345,20 +2343,13 @@ func UpdateIssueDeadline(ctx *context.Context) {
...
-	}
-
+	deadlineUnix, _ := common.ParseDeadlineDateToEndOfDay(ctx.FormString("deadline"))
 	if err := issues_model.UpdateIssueDeadline(ctx, issue, deadlineUnix, ctx.Doer); err != nil {
 		ctx.Error(http.StatusInternalServerError, "UpdateIssueDeadline", err.Error())
 		return
 	}
 
-	ctx.JSON(http.StatusCreated, api.IssueDeadline{Deadline: &deadline})
+	ctx.JSONRedirect("")
```
Comment (lunny) [inline routers/web/repo/issue.go:2352]: It seems it's not right.

Comment (wxiaoguang) [reply routers/web/repo/issue.go:2352]: I believe it is right. If you don't think so, please tell why. It is frustrating that you can not describe a problem clearly.

**Pull request merged**